### PR TITLE
Create C++ Sun Sky

### LINKED ISSUE #495: Create C++ CesiumSunSky base class
This will allow classes in C++ to call methods in CesiumSunSky without jumping through hoops like iterating through property / function names.

Blueprint logic can be kept in blueprints for ease of modification, using `BlueprintNativeEvent` so that a combination of C++ and Blueprints functionality can be used.

This will also give us an opportunity to add Android compatibility, by creating a basic DirectionalLight + Atmosphere setup using engine-provided `BP_Sky_Sphere`. It's not as realistic as the SkyAtmosphere component but it will enable other CesiumSunSky features such as time of day to be usable on Android platform.

Epic Games' response to question of Android not being supported with SkyAtmosphere component:
```
Hi, on mobile to render SkyAtmosphere you will need to enable support for Vertex Fog in the project settings and also add a Skybox mesh with a material tagged as IsSky and use the SkyAtmosphere [material] nodes to visualize the Atmosphere.

As an example for such material you can create a template map "TimeOfDay" and examine SkyDomeMesh there.

 

In 4.27 we have added on screen warning to let user know why sky atmosphere is missing on mobile. 
```

#190

### PR DESCRIPTION
**Pull request by xuelongmu**
Closes #495 

I created a new branch off of #459 

- ✅  Add android support using Engine sky material using platform-specific #ifdefs

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @xuelongmu!

* :heavy_check_mark: Signed CLA found.

Reviewers, don't forget to make sure that:

- [ ] [Cesium for Unreal Samples](https://github.com/CesiumGS/cesium-unreal-samples) works.

**Comment by xuelongmu** [comment]
I was able to fix the issue of the default skybox (`MobileSkySphere`) staying in place as the world origin shifted by checking this:
image
Now it will simply move along with the world origin as it rebases.

**Comment by xuelongmu** [comment]
The new, updated CesiumSunSky system is now a base C++ class and a child BP class called `CesiumSunSkyBP`. For backwards compatibility purposes, the old CesiumSunSky still lives in the repo (I tried marking it as `deprecated` but the scene subsequently went dark...seems like the deprecation flag broke it).

The new system will auto-adjust the radius of the SkyAtmosphere to adapt to the height and latitude of the player's current position, addressing #282.

If the `EnableMobileRendering` is checked, the CesiumSunSky will spawn a basic sky sphere, `MobileSkySphere` (based off the engine BP_Sky_Sphere). Users are encouraged to duplicate this class and add their own modifications, and then set the `SkySphereClass` variable in `CesiumSunSkyBP` to the new class. Changes in solar time, day/month, and DST will propagate to the `MobileSkySphere` as well, so Android apps can still benefit from a responsive and accurate sun.

@argallegos Would be great to get your feedback on the visual side of things.

**Comment by argallegos** [comment]
Everything looks good to me in editor. However, it looks like there are still problems with getting it to actually show up on Android devices. 

I just pushed a change to CesiumSunSky.h that tucks some variables back under advanced display and attempts to reorder some categories. I'd like to see if we can keep the variable order on the blueprint the same as the original CesiumSunSky so that our learning content remains consistent, and also for organization. However, there is apparently no way to define the order of categories in the details panel, it may depend on their order in the code but could also depend on the whims of fate. So I successfully moved the "mobile" category below the date and time information, but had to leave the "sun" category at the top because it refused to go lower than the date and time category. If anyone is able to tuck it down near mobile, that would be great, but it's probably not possible. 

I'm still looking at making a sky sphere material that more closely matches the SkyAtmosphere of CesiumSunSky, but it's not very important, so if I don't get around to it by the time this is all ready to go, that's fine.

Long story short, this has my approval visually, but looks like there's still some work to be done in function.

**Comment by xuelongmu** [comment]
From the latest testing, it looks like the Enable Mobile Rendering option is now working on Android, and is showing a `MobileSkySphere` (based off the engine `BP_Sky_Sphere`).

**Comment by shehzan10** [comment]
@xuelongmu I tested this with the provided built APK on a Samsung S21 Ultra and a Moto G8 Power. Both worked well, but it looks like the "sky" doesn't extend to the globe? See attached video.

https://user-images.githubusercontent.com/2288659/124362062-cbaaa100-dc00-11eb-9652-bc273eaa0121.mp4

**Comment by xuelongmu** [comment]
Thanks for testing @shehzan10! What you were seeing was a consequence of a sky sphere mesh that moved with the player, and also was fairly small. 

I was able to increase the sky sphere mesh size, and center it to ECEF 0,0,0, but you can no longer see the clouds. Also, when viewed from the atmosphere, you can see that it suddenly cuts from the sky sphere material to black.

image

image

**Comment by xuelongmu** [comment]
@javagl I have merged the `bp-function-library` branch (post `bp-function-library-initializations` merge) and this branch is now ready for review.

**Comment by javagl** [comment]
I tried this out to some extent, e.g. 

- Checking whether it works in general
- Changing the "Use Level Directional Light" setting
- Setting the Georeference to `None`
- Switching between normal and "Mobile Rendering" mode
- ...

and did not find any glaring issues and crashes.

Some minor, low-level comments are inlined.

Some (very) high-level comments:


- Regarding the `UpdateAtmospherePeriod`: I wondered whether it was possible to make the update "event-dependent" (and not time-dependent). Right now, when moving quickly, it can (roughly speaking) "display a 'wrong' state for 1 second". Maybe something like (hadwavingly:) `GetPlayerPawn->MovedEvent->WhenTriggered(UpdateAtmosphere)`? (We're not considering multiple pawns anyhow...)

- In the "Event Graph", the "Orient sun / directional light" block: When the "LevelDirectionalLight" is used, it sets the rotation of the directional light. Otherwise, it sets the rotation of the _Scene_. Is there a reason why it does not set the rotation of the directional light that is _contained_ in the scene? 

- More of a detail/comment: I wondered why `HandleGeoreferenceUpdated` is called when changing the location of the `CesiumSunSky` in the editor, and it was (once more) caused by the fact that changing these values calls `OnConstruction`. (They messed this one up, didn't they?). I recently tried to avoid confusion there, with a pattern like this:

      void SomeThing::OnConstruction() {
           LOG("OnConstruction");
          _updateInternalState();
      }
      void SomeThing::HandleGeoreferenceUpdated() {
           LOG("HandleGeoreferenceUpdated");
          _updateInternalState();
      }
      void SomeThing::_updateInternalState() {
           LOG("_updateInternalState, may be triggered by OnConstruction or HandleGeoreferenceUpdated");
          this->x = ...;
      }

  The point being: The georeference was *not* updated when `OnConstruction` is called - even though it might have to do the same update as in `OnConstruction` in this case. So if the same update has to be done in both cases, I extract that to a method, and let `HandleGeoreferenceUpdated` only call this (internal) method. (I don't know whether I did this 100% consistently, but I tried...)

- The fact that it was not possible to modify the location (because it is immediately overwritten with the ECEF-(0,0,0) vector) confused me at first, but since all this is about ~"something that is fixed for the globe", it certainly makes sense from a technical viewpoint


```
  
 



Unrelated to this PR (TODO: track as issue!)

When deleting the tileset and pressing CTRL+Z to undo:
```
LoginId:e814743f48dfbd95440139847b52e135
EpicAccountId:

Unhandled Exception: EXCEPTION_ACCESS_VIOLATION writing address 0x0000000000000128

UE4Editor_CesiumRuntime_Win64_DebugGame!ACesium3DTileset::Tick() [C:\Develop\CesiumUnreal\cesium-unreal-samples\Plugins\cesium-unreal\Source\CesiumRuntime\Private\Cesium3DTileset.cpp:925]
UE4Editor_Engine!AActor::TickActor() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\Actor.cpp:1093]
UE4Editor_Engine!FActorTickFunction::ExecuteTick() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\Actor.cpp:173]
UE4Editor_Engine!FTickFunctionTask::DoTask() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\TickTaskManager.cpp:284]
UE4Editor_Engine!TGraphTask<FTickFunctionTask>::ExecuteTask() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Core\Public\Async\TaskGraphInterfaces.h:886]
UE4Editor_Core!FNamedTaskThread::ProcessTasksNamedThread() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Core\Private\Async\TaskGraph.cpp:709]
UE4Editor_Core!FNamedTaskThread::ProcessTasksUntilQuit() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Core\Private\Async\TaskGraph.cpp:601]
UE4Editor_Core!FTaskGraphImplementation::WaitUntilTasksComplete() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Core\Private\Async\TaskGraph.cpp:1525]
UE4Editor_Engine!FTickTaskSequencer::ReleaseTickGroup() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\TickTaskManager.cpp:564]
UE4Editor_Engine!FTickTaskManager::RunTickGroup() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\TickTaskManager.cpp:1578]
UE4Editor_Engine!UWorld::RunTickGroup() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\LevelTick.cpp:782]
UE4Editor_Engine!UWorld::Tick() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Engine\Private\LevelTick.cpp:1576]
UE4Editor_UnrealEd!UEditorEngine::Tick() [D:\Build\++UE4\Sync\Engine\Source\Editor\UnrealEd\Private\EditorEngine.cpp:1527]
UE4Editor_UnrealEd!UUnrealEdEngine::Tick() [D:\Build\++UE4\Sync\Engine\Source\Editor\UnrealEd\Private\UnrealEdEngine.cpp:426]
UE4Editor_Win64_DebugGame!FEngineLoop::Tick() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Launch\Private\LaunchEngineLoop.cpp:4836]
UE4Editor_Win64_DebugGame!GuardedMain() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Launch\Private\Launch.cpp:169]
UE4Editor_Win64_DebugGame!GuardedMainWrapper() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Launch\Private\Windows\LaunchWindows.cpp:137]
UE4Editor_Win64_DebugGame!WinMain() [D:\Build\++UE4\Sync\Engine\Source\Runtime\Launch\Private\Windows\LaunchWindows.cpp:268]
UE4Editor_Win64_DebugGame!__scrt_common_main_seh() [d:\agent\_work\5\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288]
kernel32
ntdll
```

File: Source/CesiumRuntime/Private/CesiumGeoreference.cpp [hunk 1]
```diff
@@ -573,7 +574,7 @@ void ACesiumGeoreference::_performOriginRebasing() {
     int32 newX = clampedAdd(cameraLocation.X, originLocation.X);
     int32 newY = clampedAdd(cameraLocation.Y, originLocation.Y);
     int32 newZ = clampedAdd(cameraLocation.Z, originLocation.Z);
-    world->SetNewWorldOrigin(FIntVector(newX, newY, newZ));
+    world->RequestNewWorldOrigin(FIntVector(newX, newY, newZ));
```
Comment (javagl) [inline Source/CesiumRuntime/Private/CesiumGeoreference.cpp:577]: This looks like a subtle change, but it _might_ affect the behavior in a way that I currently cannot foresee. (I know that there was an issue in an internal project where this change solved a bug, but I'd have another close look at this at some point...)

**Review by javagl — COMMENTED** [review]
Some (minor) comments are inlined, some high-level remarks are added as a comment.

File: Source/CesiumRuntime/Private/CesiumGeoreference.cpp (hunk 1 above)
Comment (xuelongmu) [reply Source/CesiumRuntime/Private/CesiumGeoreference.cpp:577]: Thanks for making a note of that - will be good to test if this change causes any issues. I do know that the Unreal API uses RequestNewWorldOrigin in the `SetWorldOriginLocation` blueprint function.

File: Source/CesiumRuntime/Public/CesiumSunSky.h [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+      Category = Location,
+      meta = (UIMin = -12, UIMax = 14, ClampMin = -12, ClampMax = 14))
+  float TimeZone = -5.f;
+
+  UPROPERTY(
+      EditAnywhere,
+      BlueprintReadWrite,
+      Category = Location,
+      meta = (UIMin = -360, UIMax = 360, ClampMin = -360, ClampMax = 360))
+  float NorthOffset = -90.f;
```
[the remark this answers is not in the document]
Comment (xuelongmu) [reply Source/CesiumRuntime/Public/CesiumSunSky.h:115]: Added, thanks!

File: Source/CesiumRuntime/Public/CesiumSunSky.h [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+   */
+  UPROPERTY(EditDefaultsOnly, BlueprintReadWrite, Category = Sun)
+  float Azimuth = 0.f;
+
+  UPROPERTY(
+      EditAnywhere,
+      BlueprintReadWrite,
+      Category = Location,
+      meta =
+          (UIMin = -89.99, UIMax = 89.99, ClampMin = -89.99, ClampMax = 89.99))
```
[the remark this answers is not in the document]
Comment (xuelongmu) [reply Source/CesiumRuntime/Public/CesiumSunSky.h:93]: Thanks, I've removed UIMin where it's redundant. As for the `89.99`, I've copied that value from the engine SunSkyBP. I just tried testing it out with [-90, 90] and didn't notice any additional issues compared to 89.99 - however, everything gets weird near that limit anyhow. I've gone ahead and changed it.

File: Source/CesiumRuntime/Public/CesiumSunSky.h [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+   * latitude, and height. Fixes artifacts seen with the atmosphere rendering
+   * when flying high above the surface, or low to the ground in high latitudes.
+   */
+  UPROPERTY(EditAnywhere, BlueprintReadWrite, Category = Cesium)
+  bool UpdateAtmosphereAtRuntime = true;
+
+  /**
+   * How frequently the atmosphere should be updated, in seconds.
+   */
+  UPROPERTY(EditAnywhere, BlueprintReadWrite, Category = Cesium)
```
[the remark this answers is not in the document]
Comment (xuelongmu) [reply Source/CesiumRuntime/Public/CesiumSunSky.h:54]: Done

File: Source/CesiumRuntime/Private/CesiumSunSky.cpp [hunk 5]
```diff
@@ -18,6 +61,81 @@ ACesiumSunSky::ACesiumSunSky() {
...
+    float InSolarTime,
+    int32& Hour,
+    int32& Minute,
+    int32& Second) {
+  Hour = FMath::TruncToInt(InSolarTime) % 24;
+  Minute = (FMath::TruncToInt(InSolarTime - Hour) * 60) % 60;
+  
+  // Convert hours + minutes so far to seconds, subtract from InSolarTime.
+  // Not exactly sure why 0.5 is added at the end. Maybe to ensure that times
+  // on the hour (e.g. 13.0) don't get truncated down after arithmetic
```
[the remark this answers is not in the document]
Comment (xuelongmu) [reply Source/CesiumRuntime/Private/CesiumSunSky.cpp:259]: Thanks for the explanation! Makes sense to me now. I've refactored with the `RoundToInt` method.

**Comment by xuelongmu** [comment]
> * Regarding the `UpdateAtmospherePeriod`: I wondered whether it was possible to make the update "event-dependent" (and not time-dependent). Right now, when moving quickly, it can (roughly speaking) "display a 'wrong' state for 1 second". Maybe something like (hadwavingly:) `GetPlayerPawn->MovedEvent->WhenTriggered(UpdateAtmosphere)`? (We're not considering multiple pawns anyhow...)

I agree, an event-based approach would be better. Users may have their own custom pawns, so any system we create should probably use only functions available from `APawn`. Another alternative is to call `UpdateAtmospherePeriod` whenever the world origin rebases, though that can also be turned off. Unfortunately my time is limited now, so I think it would be best to create an issue to potentially improve this in the future.

> * In the "Event Graph", the "Orient sun / directional light" block: When the "LevelDirectionalLight" is used, it sets the rotation of the directional light. Otherwise, it sets the rotation of the _Scene_. Is there a reason why it does not set the rotation of the directional light that is _contained_ in the scene?

It simply sets the rotation of the root scene component in the LevelDirectionalLight, without actually requesting the `DirectionalLightComponent`. @argallegos wrote this portion, and I assume he chose to grab the root component so that the code paths would be symmetric: they would both take the root component of whatever actor the directional light component is in, and set its rotation. The reasoning behind setting the root component rotation is that it's easier during level editing to set the actor's rotation (i.e. root component rotation) rather than have to drill down to a sub-component.

> * More of a detail/comment: I wondered why `HandleGeoreferenceUpdated` is called when changing the location of the `CesiumSunSky` in the editor, and it was (once more) caused by the fact that changing these values calls `OnConstruction`. (They messed this one up, didn't they?). I recently tried to avoid confusion there, with a pattern like this...

Thanks for the tip, I copied your approach here.

> * The fact that it was not possible to modify the location (because it is immediately overwritten with the ECEF-(0,0,0) vector) confused me at first, but since all this is about ~"something that is fixed for the globe", it certainly makes sense from a technical viewpoint

It is indeed unintuitive. I added a comment in the class header file as well as the blueprint description (shown whenever you hover over an actor in the content browser) explaining this.

**Comment by xuelongmu** [comment]
Any other comments @javagl? Otherwise it would be great to get this merged in.

**Comment by javagl** [comment]
Sorry for the delay. There are some details that I'd like to review and test further (e.g. the `SetNewWorldOrigin` vs. `RequestNewWorldOrigin`), but at some point, this turns from "reviewing a PR" into "reading UE code+docs", and become less and less useful: I cannot hold this up based on the "Worst-case-'What-if?'-scenarios" that I tend to think about. I tested and reviewed it as far as reasonably possible (for me) and did not find any issues, so will merge it now.
